parse/lex.go

```diff
@@ -32,18 +36,22 @@ func New(input string) *Lexer {
 
 func (l *Lexer) run(input string) {
 
-	state := newState(input)
-	var token Token
+	state := state{
+		input: input,
+	}
 
 	for stateFn := DefaultStateFn; stateFn != nil; {
-		stateFn, state, token = stateFn(state)
-		l.tokench <- token
+		stateFn = stateFn(state, l)
 	}
 
 	close(l.tokench)
 }
 
-type stateFn func(state) (stateFn, state, Token)
+func (l *Lexer) emit(t Token) {
+	l.tokench <- t
+}
+
+type stateFn func(state, emitter) stateFn
```

Comment:
I think stateFn need detail comment(ex. working as functional programming way). because this may can be our key feature in our lexer.